federation/pkg/federation-controller/service/service_helper.go

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"time"
+	"fmt"
+
+	"k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_3"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/v1"
+	cache "k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/golang/glog"
+	"reflect"
+)
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceController) clusterServiceWorker() {
+	fedClient := sc.federationClient
+	for clusterName, clusterCache := range sc.clusterCache.clientMap{
+		go func() {
+			for {
+				key, quit := clusterCache.serviceQueue.Get()
+				// update service cache
+				if quit {
+					return
+				}
+				defer clusterCache.serviceQueue.Done(key)
+				err := sc.clusterCache.syncService(key.(string), clusterName, clusterCache, sc.serviceCache, fedClient)
+				if err != nil {
+					glog.V(2).Infof("failed to sync service: %+v", err)
+					//return err
```

Comment:
I think that you need to prevent the deferred Done() from being called here, as you're not done for this key?
Perhaps only call Done() if err == nil?  On the other hand, how to avoid stalling the queue for a service that consistently fails to sync?